consensus/istanbul/backend/handler.go

```
@@ -290,3 +298,171 @@ func (sb *Backend) UnregisterPeer(peer consensus.Peer, isProxiedPeer bool) {
+}
+
+// shouldSendValidatorProof determines if this node should send a
+// validator proof revealing its address to a peer
+func (sb *Backend) shouldSendValidatorProof(peer consensus.Peer) (bool, error) {
+	var validatorAddress common.Address
+	if sb.config.Proxy {
+		validatorAddress = sb.config.ProxiedValidatorAddress
+	} else {
+		validatorAddress = sb.Address()
+	}
+	// Check to see if this node is a validator
+	regAndActiveVals, err := sb.retrieveRegisteredAndElectedValidators()
+	if err != nil {
+		return false, err
+	}
+	if !regAndActiveVals[validatorAddress] {
+		return false, nil
+	}
+	// If the peer is not a known validator, we do not want to give up extra
+	// information about ourselves
+	if _, err := sb.valEnodeTable.GetAddressFromNodeID(peer.Node().ID()); err != nil {
+		return false, nil
+	}
+
+	return true, nil
+}
+
+// readValidatorProofMessage reads a validator proof message as a part of the
+// istanbul handshake. Returns if the peer is a validator or if an error occurred.
+func (sb *Backend) readValidatorProofMessage(peer consensus.Peer) (bool, error) {
+	peerMsg, err := peer.ReadMsg()
+	if err != nil {
+		return false, err
+	}
+
+	var payload []byte
+	if err := peerMsg.Decode(&payload); err != nil {
+		return false, err
+	}
+
+	var msg istanbul.Message
+	err = msg.FromPayload(payload, sb.verifyValidatorProofMessage)
+	if err != nil {
+		return false, err
+	}
+	// If the Msg is empty, the peer has decided not to reveal its info
+	if len(msg.Msg) == 0 {
+		return false, nil
+	}
+
+	var directAnnounce directAnnounce
+	err = rlp.DecodeBytes(msg.Msg, &directAnnounce)
+	if err != nil {
+		return false, err
+	}
+
+	// Check if the peer is within the registered/elected valset
+	regAndActiveVals, err := sb.retrieveRegisteredAndElectedValidators()
+	if err != nil {
+		sb.logger.Trace("Error in retrieving registered/elected valset", "err", err)
+		return false, err
+	}
+	if !regAndActiveVals[msg.Address] {
+		return false, nil
```

nit: Should we return an error here? Or log something? IIUC, with everyone running the proper code, we should never, or almost never, reach this line